### ISSUE TITLE
Proper way to measure the latency of a torch-xla script

### ISSUE THREAD

**Issue by wangshangsam** [pytorch/xla#2492]
## ❓ What's the proper way of measuring the training loop runtime of a `torch-xla` script?

Currently I have a training loop that looks something like this:
```python
timing = {'epoch': [], 'epoch_start': [], 'epoch_stop': []}
for epoch in range(epochs):
  timing['epoch'].append(epoch)
  timing['epoch_start'].append(time.time())
  for i, data in enumerate(dataloader, 0):
    # forward pass
    # backward pass
    # xm.optimizer_step(optimizer, barrier=True)
  timing['epoch_stop'].append(time.time())
  print('Epoch {} took {} s!'.format(
      epoch,
      timing['epoch_stop'][-1] - timing['epoch_start'][-1],
  ))
```
When I try to measure how long each epoch takes via the above code for 1) one model 2) the same model/code with 2X batch size, for some reason, I'm observing "superlinear scaling", that is, 2) takes much less time than 1). I understand my model could be very un-optimized in many ways, but I don't quite see how "superlinear scaling" could even be possible. Therefore, I'm wondering if measuring runtime in the above approach is actually correct?

Thanks a lot!

@eric-zheng

PS: Is xprof publically usable for `torch-xla` now? cuz otherwise, I'm really in the dark to debug a problem such as above ;)

**Comment by jysohn23** [comment]
TPUs have a large MXU, which makes it very efficient to do large matrix multiplications on it. Typically to maximize utilization you want to fit the max batch size that can fit on the TPU before the HBM OOMs. And yes, in case the TPU was being underutilized with small batch size, superlinear scaling would make sense since you'd be improving TPU utilization.

Yes, your benchmarking looks correct as you're inserting a barrier to the step. However, to even further improve utilization, you want to you a `pl.MpDeviceLoader` as shown in our https://github.com/pytorch/xla/blob/master/API_GUIDE.md This loader pipelines data transfer with graph building and execution (in case you're not already using this loader).

Also, yes `capture_tpu_profile` tool is usable for pytorch/xla, but there may be some rough edges as we haven't spent too much time making sure all functionality is compatible, but you should be able to see at least utilization metrics.

**Comment by wangshangsam** [comment]
Thanks a lot for reminding me of `pl.MpDeviceLoader`! However, I still have a question:

> TPUs have a large MXU, which makes it very efficient to do large matrix multiplications on it. Typically to maximize utilization you want to fit the max batch size that can fit on the TPU before the HBM OOMs.

This I understand, but I'm a little confused about
> in case the TPU was being underutilized with small batch size, superlinear scaling would make sense

Because "superlinear scaling" means, on the same MXU, a larger GEMM has shorter latency than a smaller GEMM, and I don't quite see how that is possible in the hardware (unless this is intentionally designed / a hardware feature). Does TPU dynamically control the power and/or clock frequency based on the GEMM size or something like this?

**Comment by jysohn23** [comment]
There may be many causes, but the main one that comes to mind is the possibility that in the smaller batch size case, the tensors shapes were being padded to fit well in TPU memory, which goes away once your batch size increases enough not to need any padding so we no longer waste FLOPs. For more info try reading: https://cloud.google.com/tpu/docs/troubleshooting#batch-too-small
